### ISSUE TITLE
256 byte transfer over Tube hangs on Tube release call

### ISSUE THREAD

**Issue by kieranhj** [stardot/b-em#30]
Boot the attached ssd [tube-bug.zip](https://github.com/stardot/b-em/files/1450767/tube-bug.zip) in Master Turbo mode. In MODE 7 you should see a message on screen, in any other mode a flash of green raster and then some crap left in the screen buffer.

This is copying 256 bytes from the parasite RAM over to the host RAM using the 256 byte "fast" transfer protocol - see AUG page 338 for details.

Alas on b-em (and jsbeeb) the Tube code hangs during the Tube release call waiting for a byte from the parasite that never arrives. This has been verified to work on my actual Master Turbo hardware.

**Comment by SteveFosdick** [comment]
Looking at this with the debugger in B-Em on the tube processor it looks like the transfer never actually happens.  I am referring to the source code of the tube code at http://mdfs.net/Software/Tube/6502/Client65.src.  From when the *HOST command is issued, the tube ROM sends then command and then sits in a loop polling FEFA (S2).  While still in that loop it receives an IRQ to initiate the data transfer, receives code, sender ID, address and sync byte and then sits in a loop examining FEFC (S3) waiting for R3 to become free which never happens.

**Comment by SteveFosdick** [comment]
Looking a little further this looks like a tube ROM bug rather than a B-Em bug.  In the source code for the Tube Client ROM on JGH's site as mentioned in the previous comment the code to send 256 bytes from parasite to host starts with:

```
\ Send 256 bytes to Tube via R3
\ -----------------------------
LDY #&00
.LFDCF
LDA TubeS3:AND #&80:BPL LFDCF:\ Wait for Tube R3 `free
```
According to the documentation at http://mdfs.net/Docs/Comp/BBC/Hardware/TubePorts bit 7 of S3, as tested by AND #&80 and BPL is Rx ready yet we are waiting to transmit!  Also later down the same source file the code for the parasite to receive 256 bytes from the host starts with the same instructions, commented differently:

```
\ Read 256 bytes from Tube via R3
\ -------------------------------
.LFDEC
LDY #&00
.LFDEE
LDA TubeS3:AND #&80:BPL LFDEE:\ Wait for Tube R3 data present
```
so as a test I patched the first subroutine and replaced the polling loop waiting to send with:

```
.loop
    LDA `S3
    AND #&40
    BEQ loop`
```
and it worked correctly.

**Comment by SteveFosdick** [comment]
You could try the attached 6502Tube.rom in place of the original in the roms/tube directory. I also attach the modified source code.  If you fancy assembling it yourself the source seems to be in BBC BASIC format but not quite compatible with BeebAsm so I thought I would just *EXEC it in B-Em - I needed to resort to 2nd processor and stripping the comments first on Linux before transferring the file.  I think JGH must use a version of Basic that runs on RiscOS.
[Client65.txt](https://github.com/stardot/b-em/files/1477219/Client65.txt)
[Client65.zip](https://github.com/stardot/b-em/files/1477223/Client65.zip)

Just to clarify the file naming, Client65 is the name on JGH's website but it is 6502Tube.rom in the B-Em distribution.  I had to put the ROM file in a ZIP archive for GitHub to let me upload it.

**Comment by hoglet67** [comment]
Steve,

Status register 3 works differently to the other status registers.

Bit 7 means "action required" and can be triggered be either the the P->H fifo going empty or the H->P fifo going not empty.

From a hardware perspective, it closely matches the logic of the NMI output (but without the masking).

For more details, see page 14:
http://mdfs.net/Info/Comp/Acorn/AppNotes/004.pdf#page=14

The bit is called "N", rather than "A3".

So I think fixing the tube ROM is not correct here.... as it's not reflecting the real hardware.

Dave

**Comment by hoglet67** [comment]
The PiTubeDirect tube emulation started life as the tube.c code from B-Em.

We had the same bug, and the fix was this commit:
https://github.com/hoglet67/PiTubeDirect/commit/f641dd126df7672033b7cff400189d7093beaee3#diff-7d00448e1b91c8e81cb53c638f2d30ff

It might be worth trying to make the same change to tube.c and see if that helps.

**Comment by kieranhj** [comment]
Hey Steve - thanks for investigating this issue. And Dave - thank you for fixing this issue! I just tried the changes on my local build of b-em and the test app completes as expected (ie. same as I saw on my real Master Turbo machine.)

This is great news as it means I can start exploring the potential of using the Tube for interesting Beeb fx again - if only I didn’t have PoP to finish. Too many things to explore and so little time!!
